### PR TITLE
ARROW-17600: [Go] Implement Casting for Nested types

### PR DESCRIPTION
**Pull request by zeroshade**

### CODE REVIEW

File: ci/scripts/go_test.sh [hunk 1]
```diff
@@ -19,12 +19,20 @@
 
 set -ex
 
+# simplistic semver comparison
+verlte() {
+    [ "$1" = "`echo -e "$1\n$2" | sort -V | head -n1`" ]
```
Comment (lidavidm) [inline ci/scripts/go_test.sh:24]: :rofl:

File: go/arrow/array/struct.go [hunk 2]
```diff
@@ -36,6 +36,37 @@ type Struct struct {
...
 }
 
+// NewStructArray constructs a new Struct Array out of the columns passed
+// in and the field names. The length of all cols must be the same and
+// there should be the same number of columns as names.
+func NewStructArray(cols []arrow.Array, names []string) (*Struct, error) {
+	return NewStructArrayWithNulls(cols, names, nil, 0, 0)
+}
+
+func NewStructArrayWithNulls(cols []arrow.Array, names []string, nullBitmap *memory.Buffer, nullCount int, offset int) (*Struct, error) {
```
Comment (lidavidm) [inline go/arrow/array/struct.go:46]: Should this have a docstring too?

File: go/arrow/datatype.go [hunk 3]
```diff
@@ -192,6 +192,7 @@ type FixedWidthDataType interface {
 	DataType
 	// BitWidth returns the number of bits required to store a single element of this data type in memory.
 	BitWidth() int
+	Bytes() int
```
Comment (lidavidm) [inline go/arrow/datatype.go:195]: Probably deserves a docstring too?

**Review by lidavidm — APPROVED** [review]

File: go/arrow/array/struct.go (hunk 2 above)
Comment (zeroshade) [reply go/arrow/array/struct.go:49]: added

File: go/arrow/datatype.go (hunk 3 above)
Comment (zeroshade) [reply go/arrow/datatype.go:196]: added

**Pull request merged**